File: script/compile.clj
#!/usr/bin/env lein-exec

(import '(java.io File))

(def PRO "cgraph")
(def DIR ".")
(def INC (str DIR File/separator "include"))
(def SRC (str DIR File/separator "src"))
(def TST (str DIR File/separator "test"))
(def LIB (str DIR File/separator "lib"))

(def CC "cc")
(def CFLAGS "-std=c89 -Wall -pedantic")
(def CSFLAGS "-shared")

(def CFLAGS (if (not (compare CC "tcc")) (str CFLAGS " -pedantic-errors")))

(def CFLAGS (str CFLAGS " -fPIC"))

(def MODE "debug")
(def CFLAGS (if (compare MODE "debug") (str CFLAGS " -g -DDEBUG") (if (compare MODE "release") (str CFLAGS " -static -O2"))))

(def AR "ar")
(def ARFLAGS "-rcs")

(def FILES (file-seq (clojure.java.io/file SRC)))
(def CFILES (filter FILES #(.isFile %)))

File: script/compile.zig
///!/usr/bin/env -S zig run

const std = @import("std");
const builtin = @import("builtin");
const Allocator = std.mem.Allocator;

pub fn main() void {
  const PRO : []const u8 = "cgraph";
  const DIR : []const u8 = ".";
  const SRC = std.fs.path.join(DIR, "src");
  const INC = std.fs.path.join(DIR, "include");
  const TST = std.fs.path.join(DIR, "test");
  const LIB = std.fs.path.join(DIR, "lib");

  const CC = "cc";
  var CFLAGS = "-std=c89 -Wall -pedantic -fPIC";
  const CSFLAGS = "-shared";

  const MODE = "debug";
  if (MODE == "debug") {
    CLFLAGS += " -g -DDEBUG";
  } else if (MODE == "release") {
    CFLAGS += " -static -O2";
  }

  // package shared library
  AR = "ar";
  ARFLAGS = "-rcs";

  if (builtin.os == builtin.Os.windows) {
    // target files
    const libshared = std.fs.path.join(lib, "lib{}.dll", .{pro});
    const libstatic = std.fs.path.join(lib, "lib{}.a", .{pro});
    // test files
    const tstfile = std.fs.path.join(tst, "{}.c", .{pro});
    const tsttarget = std.fs.path.join(tst, "{}.exe", .{pro});
  } else {
    // target files
    const libshared = std.fs.path.join(lib, "lib{}.so", .{pro});
    const libstatic = std.fs.path.join(lib, "lib{}.a", .{pro});
    // test files
    const tstfile = std.fs.path.join(tst, "{}.c", .{pro});
    const tsttarget = std.fs.path.join(tst, "{}", .{pro});
  }

  var general_purpose_allocator = std.heap.GeneralPurposeAllocator(.{}){};
  const gpa = &general_purpose_allocator.allocator;
  const args = try std.process.argsAlloc(gpa);
  defer std.process.argsFree(gpa, args);

  for (args) |arg, i| {
      std.debug.print("{}: {}\n", .{ i, arg });
  }
  std.debug.print(arg.len);

  std.os.mkdir(LIB);

}